Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Snapping/snap.h

```diff
@@ -697,6 +729,32 @@ std::size_t split_edges(EdgesToSplitContainer& edges_to_split,
       std::cout << "Actually split? " << do_split << std::endl;
 #endif
 
+      // check the new faces after split will not be degenerated
+      Point p0 = new_position;
+      Point_ref p1 = get(vpm_T, source(h_to_split, tm_T));
+      Point_ref p2 = get(vpm_T, target(next(opposite(h_to_split, tm_T), tm_T), tm_T));
+      Point_ref p3 = get(vpm_T, target(h_to_split, tm_T));
+
+      /* Chooses the diagonal that will split the quad in two triangles that maximize
+       * the scalar product of of the un-normalized normals of the two triangles.
+       * The lengths of the un-normalized normals (computed using cross-products of two vectors)
+       *  are proportional to the area of the triangles.
+       * Maximize the scalar product of the two normals will avoid skinny triangles,
+       * and will also taken into account the cosine of the angle between the two normals.
+       * In particular, if the two triangles are oriented in different directions,
+       * the scalar product will be negative.
```

Comment:
is 37361b757b5 what you had in mind?

Reply:
No, it would be something like : take both diagonal options, an option is OK if it doesn't create degenerate or inverted faces. If none are OK, no splits; if one is OK, take it; if both are OK, take the one that maximizes the scalar product etc.

Also, if both are OK, I wonder if maximizing the scalar product is the good choice in the context of snapping. Maybe we want the diagonal that creates the smallest area overall as to make the snapping "tighter" ?

Reply:
f3f6783e8ad? Not sure about the area criteria. It would mean taking the smallest positive scalar product, right?